lib/src/space/grounding.rs

```diff
@@ -47,7 +50,25 @@ impl<'a> Iterator for GroundingSpaceIter<'a> {
     }
 }
 
-fn atom_to_trie_key(atom: &Atom) -> TrieKey<SymbolAtom> {
+impl Serializer for DefaultHasher { // there are much speedier hashers, but I didn't want to introduce a dependency
+    fn serialize_bool(&mut self, _v: bool) -> serial::Result { Ok(self.write_u8(_v as u8)) }
+    fn serialize_i64(&mut self, _v: i64) -> serial::Result { Ok(self.write_i64(_v)) }
+    fn serialize_f64(&mut self, _v: f64) -> serial::Result { Ok(self.write_u64(_v as u64)) }
+}
+
+impl Serializer for String { // for debugging
+    fn serialize_bool(&mut self, _v: bool) -> serial::Result { Ok(self.push_str(&*_v.to_string())) }
+    fn serialize_i64(&mut self, _v: i64) -> serial::Result { Ok(self.push_str(&*_v.to_string())) }
+    fn serialize_f64(&mut self, _v: f64) -> serial::Result { Ok(self.push_str(&*_v.to_string())) }
+}
+
+impl Serializer for Vec<u8> { // for speed, but is technically unsafe because not a valid utf-8 string
```

Comment:
There is nothing unsafe about pushing data onto a vec of bytes.  The unsafe thing would be if you then interpreted the bytes as a str / String later on, but I don't see you doing that.

So I think this comment is unnecessarily alarming.